omx/gstomxcamera.c

```diff
@@ -0,0 +1,1214 @@
+              }
+            }
+          }
+        }
+
+        caps = gst_caps_copy_nth (icaps, best);
+        gst_caps_unref (icaps);
+      } else {
+        caps = icaps;
+      }
+    }
+    gst_caps_unref (thiscaps);
+  } else {
+    /* no peer or peer have ANY caps, work with our own caps then */
+    caps = thiscaps;
+  }
+  if (peercaps)
+    gst_caps_unref (peercaps);
+  if (caps) {
+    caps = gst_caps_truncate (caps);
+
+    /* now fixate */
+    if (!gst_caps_is_empty (caps)) {
+      caps = gst_omx_camera_fixate (basesrc, caps);
+      GST_DEBUG_OBJECT (basesrc, "fixated to: %" GST_PTR_FORMAT, caps);
+
+      if (gst_caps_is_any (caps)) {
+        /* hmm, still anything, so element can do anything and
+         * nego is not needed */
+        result = TRUE;
+      } else if (gst_caps_is_fixed (caps)) {
+        /* yay, fixed caps, use those then */
+        result = gst_base_src_set_caps (basesrc, caps);
+      }
+    }
+    gst_caps_unref (caps);
+  }
+  return result;
+
+no_nego_needed:
+  {
+    GST_DEBUG_OBJECT (basesrc, "no negotiation needed");
+    if (thiscaps)
+      gst_caps_unref (thiscaps);
+    return TRUE;
+  }
+}
+
+OMX_COLOR_FORMATTYPE
+gst_omx_camera_get_color_format (GstVideoFormat format)
+{
+  OMX_COLOR_FORMATTYPE omx_format;
+
+  switch (format) {
+    case GST_VIDEO_FORMAT_YUY2:
+      omx_format = OMX_COLOR_FormatYCbYCr;
+      break;
+    case GST_VIDEO_FORMAT_I420:
+      omx_format = OMX_COLOR_FormatYUV420Planar;
+      break;
+    case GST_VIDEO_FORMAT_NV12:
+      omx_format = OMX_COLOR_FormatYUV420SemiPlanar;
+      break;
+    default:
+      omx_format = OMX_COLOR_FormatUnused;
+      break;
+  }
+  return omx_format;
+}
+
+gint
+gst_omx_camera_get_buffer_size (GstVideoFormat format, gint stride, gint height)
+{
+  gint buffer_size;
+
+  switch (format) {
+    case GST_VIDEO_FORMAT_YUY2:
+      buffer_size = stride * height;
+      break;
+    case GST_VIDEO_FORMAT_I420:
+      buffer_size = stride * height + 2 * ((stride >> 1) * ((height + 1) >> 2));
+      break;
+    case GST_VIDEO_FORMAT_NV12:
+      buffer_size = (stride * height * 3) >> 1;
+      break;
+    default:
+      buffer_size = 0;
+      break;
+  }
+  return buffer_size;
+}
+
+
+static gboolean
+gst_omx_camera_drain (GstOMXCamera * self)
+{
+  if (self->outport)
+    gst_omx_port_set_flushing (self->outport, 5 * GST_SECOND, TRUE);
+
+  if (!gst_omx_camera_stop (GST_BASE_SRC (self)))
+    return FALSE;
+
+  if (!gst_omx_camera_shutdown (self))
+    return FALSE;
+  GST_DEBUG_OBJECT (self, "OMX camera drained and disabled");
+
+  return TRUE;
+}
+
+static gboolean
+gst_omx_camera_set_format (GstOMXCamera * self, GstCaps * caps,
+    GstVideoInfo * info)
+{
+
+  gboolean needs_disable = FALSE;
+  OMX_PARAM_PORTDEFINITIONTYPE port_def;
+  OMX_PARAM_BUFFER_MEMORYTYPE mem_type;
+  OMX_PARAM_VFCC_HWPORT_PROPERTIES hw_port_param;
+  OMX_PARAM_VFCC_HWPORT_ID hw_port;
+  OMX_ERRORTYPE err;
+  GstStructure *config;
+
```

Comment:
In C/C++ we should check for every pointer to be valid. We call these pointer guards. The glib/gst pattern is to use the `g_return_val_if_fail` family. You should use these in every method where you receive a pointer.

Reply:
Ready for review: 1586b6c19e45d944964548c2177f389f3e98f6a9